Apps/Admin/Server/Services/AgentAccessService.cs

```diff
@@ -0,0 +1,202 @@
+        /// </summary>
+        /// <param name="authDelegate">The OAuth2 authentication service.</param>
+        /// <param name="keycloakAdminApi">The keycloak api to access identity access.</param>
+        /// <param name="configuration">The configuration to use.</param>
+        /// <param name="logger">Injected Logger Provider.</param>
+        public AgentAccessService(
+            IAuthenticationDelegate authDelegate,
+            IKeycloakAdminApi keycloakAdminApi,
+            ILogger<AgentAccessService> logger,
+            IConfiguration configuration)
+        {
+            this.authDelegate = authDelegate;
+            this.keycloakAdminApi = keycloakAdminApi;
+            this.logger = logger;
+            _ = configuration;
+            (this.tokenUri, this.tokenRequest) = this.authDelegate.GetClientCredentialsAuth(AuthConfigSectionName);
+        }
+
+        /// <inheritdoc/>
+        [SuppressMessage("Globalization", "CA1308:Normalize strings to uppercase", Justification = "Not making security decisions based on the result of the normalization")]
+        public async Task<AdminAgent> ProvisionAgentAccessAsync(AdminAgent agent)
+        {
+            JwtModel jwtModel = this.authDelegate.AuthenticateAsSystem(this.tokenUri, this.tokenRequest);
+
+            agent.Username = agent.Username.ToLowerInvariant();
+            string userName = agent.Username + "@" + EnumUtility.ToEnumString<KeycloakIdentityProvider>(agent.IdentityProvider, true);
+            IEnumerable<RoleRepresentation> roles = await this.GetRoleRepresentationsAsync(agent.Roles).ConfigureAwait(true);
+
+            UserRepresentation user = new()
+            {
+                Username = userName,
+                Enabled = true,
+            };
+
+            try
+            {
+                await this.keycloakAdminApi.AddUserAsync(user, jwtModel.AccessToken).ConfigureAwait(true);
+            }
+            catch (Exception e) when (e is ApiException)
+            {
+                if (e.Message.Contains("409", StringComparison.InvariantCulture))
+                {
+                    this.logger.LogWarning(ErrorMessages.KeycloakUserAlreadyExists);
+                    throw new ProblemDetailsException(ExceptionUtility.CreateProblemDetails(ErrorMessages.KeycloakUserAlreadyExists, HttpStatusCode.Conflict, nameof(AgentAccessService)));
+                }
+
+                this.logger.LogError(e, "Keycloak API call failed");
+                throw;
+            }
+
+            List<UserRepresentation> createdUser = await this.keycloakAdminApi.GetUserAsync(userName, jwtModel.AccessToken).ConfigureAwait(true);
+            await this.keycloakAdminApi.AddUserRolesAsync(createdUser.First().UserId.ToString(), roles, jwtModel.AccessToken).ConfigureAwait(true);
+
+            return agent;
+        }
+
+        /// <inheritdoc/>
+        public async Task<IEnumerable<AdminAgent>> GetAgentsAsync(string queryString, int? resultLimit = 25)
```

Comment:
Looks like a security risk, does it make sense to pass user input directly from the controller to KC?

Reply:
I think this is safe for a few reasons:
1) The application audience generally has equivalent admin access in Keycloak
2) The sanitation of the input should be handled by the service provider (Keycloak) as it is their consumable API.
3) What would make it safe?  We could strip out characters but which ones are invalid for Keycloak:
<img width="964" alt="image" src="https://user-images.githubusercontent.com/51342761/212422791-d9e2a147-e0c6-4409-bebf-189ddb3db0cd.png">

Reply:
Ahh, I traced the actual Refit map, it's not a query string, it's a search param, then yes, not a risk. I'd recommend renaming queryString=>searchCriteria to make it clear this is not just forwarding a query string to the API call

Reply:
I still think it is worth renaming queryString=>searchString